src/components/clusterexplorer/explorer.ts

```diff
@@ -156,7 +202,86 @@ export class KubernetesExplorer implements vscode.TreeDataProvider<ClusterExplor
         // and always queue a refresh.
         //
         // These are pretty niche cases, so I'm not too worried if they aren't perfect.
-        sleep(50).then(() => refreshExplorer());
+        if (!node) {
+            sleep(50).then(() => refreshExplorer());
+        } else {
+            const currentNodeId = this.getWatchId(node as ClusterExplorerNode);
+            if (!currentNodeId) {
+                return;
+            }
+            // In the case where many requests of updating are received in a short amount of time
+            // the tree is not refreshed for every change but once after a while.
+            // Every call resets a timer which trigger the tree refresh
+            clearTimeout(this.refreshTimer);
+            // check if element already is in refreshqueue before pushing it
+            const hasNode = this.refreshQueue.find((nodeQueue) => {
+                const nodeQueueid = this.getWatchId(nodeQueue);
+                return nodeQueueid && currentNodeId === nodeQueueid;
+            });
+            if (!hasNode) {
+                this.refreshQueue.push(node as ClusterExplorerNode);
+            }
+            this.refreshTimer = setTimeout(() => {
+                this.refreshQueue.splice(0).forEach((_node) => this.refresh(_node));
+            }, 500);
+        }
+    }
+
+    private onElementCollapsed(e: vscode.TreeViewExpansionEvent<ClusterExplorerNode>) {
+        this.collapse(e.element);
+	}
+
+	private onElementExpanded(e: vscode.TreeViewExpansionEvent<ClusterExplorerNode>) {
+        this.expand(e.element);
+    }
+
+    private expand(node: ClusterExplorerNode) {
+        const resourcesToWatch = getResourcesToBeWatched();
+        const watchId = this.getWatchId(node);
+        const treeItem = node.getTreeItem();
+        providerResult.transform(treeItem, (ti) => {
+            if (ti.collapsibleState === vscode.TreeItemCollapsibleState.Collapsed) {
```

Comment:
No you're right. I added them to maintain consistency on `collapsibleState`, but it isn't used for the moment

Reply:
I'd recommend removing them if they aren't used.  Otherwise we'll forget that they're not needed and will become things we are afraid to touch because we don't know why they're needed!